app/com/linkedin/drelephant/mapreduce/MapReduceFSFetcherHadoop2.java

```diff
@@ -0,0 +1,305 @@
+    this._fetcherConfigurationData = fetcherConfData;
+
+    Configuration conf = new Configuration();
+    this._fs = FileSystem.get(conf);
+    this._historyLocation = conf.get("mapreduce.jobhistory.done-dir");
+    this._intermediateHistoryLocation = conf.get("mapreduce.jobhistory.intermediate-done-dir");
+    logger.info("Intermediate history dir: " + _intermediateHistoryLocation);
+    logger.info("History done dir: " + _historyLocation);
+  }
+
+  /**
+   * The location of a job history file is in format: {done_dir}/yyyy/mm/dd/{serialPart}.
+   * yyyy/mm/dd is the year, month and date of the finish time.
+   * serialPart is the first 6 digits of the serial number considering it as a 9 digits number.
+   * PS: The serial number is the last part of an app id.
+   * <p>
+   * For example, if appId = application_1461566847127_84624, then serial number is 84624.
+   * Consider it as a 9 digits number, serial number is 000084624. So the corresponding
+   * serialPart is 000084. If this application finish at 2016-5-30, its history file will locate
+   * at {done_dir}/2016/05/30/000084
+   */
+  private String getHistoryDir(AnalyticJob job) {
+    // generate the date part
+    Calendar timestamp = Calendar.getInstance();
+    timestamp.setTimeInMillis(job.getFinishTime());
+    String datePart = String.format(TIMESTAMP_DIR_FORMAT,
+            timestamp.get(Calendar.YEAR),
+            timestamp.get(Calendar.MONTH) + 1,
+            timestamp.get(Calendar.DAY_OF_MONTH));
+
+    // generate the serial part
+    String appId = job.getAppId();
+    int serialNumber = Integer.parseInt(appId.substring(appId.lastIndexOf('_') + 1));
+    String serialPart = String.format("%09d", serialNumber)
+            .substring(0, SERIAL_NUMBER_DIRECTORY_DIGITS);
+
+    return _historyLocation + File.separator + datePart + File.separator + serialPart + File.separator;
+  }
+
+  private DataFiles getHistoryFiles(AnalyticJob job) throws IOException {
+    String jobId = Utils.getJobIdFromApplicationId(job.getAppId());
+    String jobConfPath = null;
+    String jobHistPath = null;
+
+    // Search files in done dir
+    String jobHistoryDirPath = getHistoryDir(job);
+    RemoteIterator<LocatedFileStatus> it;
+    try {
+      it = _fs.listFiles(new Path(jobHistoryDirPath), false);
+    } catch (FileNotFoundException e) {
+      logger.error("Can't find job info of " + jobId + ": directory " + jobHistoryDirPath + " not found");
+      return null;
+    }
+    while (it.hasNext() && (jobConfPath == null || jobHistPath == null)) {
+      String name = it.next().getPath().getName();
+      if (name.contains(jobId)) {
+        if (name.endsWith("_conf.xml")) {
+          jobConfPath = jobHistoryDirPath + name;
+        } else if (name.endsWith(".jhist")) {
+          jobHistPath = jobHistoryDirPath + name;
+        }
+      }
+    }
+
+    // If some files is missing, search in the intermediate-done-dir in case that HistoryServer has not
```

Comment:
If some files **_are_** missing, search in the intermediate-done-dir in case the HistoryServer has not **_yet_** move**_d_** them into the done-dir.`